### ISSUE TITLE
[Apache] Split out span creation from baggage injection for interceptors

### ISSUE THREAD

**Issue by dray92** [jaegertracing/jaeger-client-java#240]
I think we need to split out the logic in the Request Interceptor. [`Span creation`](https://github.com/uber/jaeger-client-java/blob/master/jaeger-apachehttpclient/src/main/java/com/uber/jaeger/httpclient/TracingRequestInterceptor.java#L69) and [`injection`](https://github.com/uber/jaeger-client-java/blob/master/jaeger-apachehttpclient/src/main/java/com/uber/jaeger/httpclient/TracingRequestInterceptor.java#L81) might be a little **too** coupled. 

Suggested solution: 
 - Have one interceptor tasked with starting span -- will be added to HttpProcessor chain using [`addInterceptorFirst(HttpRequestInterceptor ..)`](https://hc.apache.org/httpcomponents-asyncclient-dev/httpasyncclient/apidocs/org/apache/http/impl/nio/client/HttpAsyncClientBuilder.html#addInterceptorFirst(org.apache.http.HttpRequestInterceptor)).
Say, `JaegerSpanCreationRequestInterceptor`
 - Have second interceptor that is tasked with span baggage injection -- will be added to the **end** of the HttpProcessor chain using [`addInterceptorLast(HttpRequestInterceptor ..)`](https://hc.apache.org/httpcomponents-asyncclient-dev/httpasyncclient/apidocs/org/apache/http/impl/nio/client/HttpAsyncClientBuilder.html#addInterceptorLast(org.apache.http.HttpRequestInterceptor)).
Say, `JaegerSpanInjectionRequestInterceptor`

This can be default behavior for something like [`TracingInterceptors.java`](https://github.com/uber/jaeger-client-java/blob/master/jaeger-apachehttpclient/src/main/java/com/uber/jaeger/httpclient/TracingInterceptors.java) - but we have to find a way to enable some interceptors to come in between the span creation and injection stages.

Cleaner way below:

As a consumer, I can do something like 
```
HttpProcessor inHttpProc = new ImmutableHttpProcessor(
    new HttpRequestInterceptor[] {
        new JaegerSpanCreationRequestInterceptor(tracer),
        new PreprocessingRequestInterceptor(),
        new AuthRequestInterceptor(),
        new PostprocessingRequestInterceptor(),
        new JaegerSpanInjectionRequestInterceptor(tracer)
    }
);
```

Then, client initialization can be something like,
```
CloseableHttpAsyncClient client = HttpAsyncClients.custom()
                                                  .setHttpProcessor(inHttpProc)
                                                  .build();
```

**Comment by black-adder** [comment]
Maybe I'm missing something obvious, but can't you create a `AuthRequestInterceptor extends TracingRequestInterceptor` which overwrites the onStartSpan function which allows you to update the span context before the injection?

**Comment by dray92** [comment]
Sure, that's a possibility for when just one `Interceptor` needs to be added. But as soon as they are multiple, doesn't it start to become extremely convoluted?

**Comment by black-adder** [comment]
Is that a real use case that you have?

**Comment by dray92** [comment]
Let's discuss about this offline

**Comment by vprithvi** [comment]
I'm not sure why this needs to be done, or why any business specific use case needs to be between span creation and baggage injection.

**Comment by jkanywhere** [comment]
@vprithvi This needs to be done so that the service to service authentication injector can add an authentication token as baggage to a span after the span is created, but before the baggage is injected into the request headers.

**Comment by dray92** [comment]
Fixed by PR #248
